### PR TITLE
Options for Stagger model loading for low memory systems

### PR DESCRIPTION
**Pull request by jjhursey**
* `--stagger_load` : (default: `0` off) Stagger model loading to avoid OOM issues on the host
 * `--stagger_update_lazyhandle` : (default: `0` off) Stagger update_lazyhandle to avoid OOM issues on the host
 * `--dist_timeout` : (default: either `10` for NCCL or `30` for others set by PyTorch) torch distributed timeout in minutes

### CODE REVIEW

**Comment by thanh-lam** [comment]
For testing purposes: What are differences between option `2` and `1`?
@jjhursey Can you add some descriptions for these options?

**Comment by jjhursey** [comment]
> For testing purposes: What are differences between option `2` and `1`?

We can improve the help text, but this is what it reports now:
```
    "--stagger_load",
    help="Stagger model loading to avoid OOM issues on the host"

    "--stagger_update_lazyhandle",
    help="Stagger update_lazyhandle to avoid OOM issues on the host"
```

So two different options to stagger processing of two different sections of the compile phase in case you have a need to set different values for each part.

The default value is `0` which means skip the staggering and let all of the processes enter this section at the same time. This results in the maximum amount of memory utilization, but also (generally) the fastest time through the compile phase.

Setting this value to `>0` defines how many concurrent processes are allowed to be in that phase at the same time. So if you run with `16` processes and set the value to `2` then only `2` processes at a time will be performing the compile phase. Once the first set of 2 are done then the next set of `2` will proceed and so on. All processes that have completed the compile will wait for the others to finish before moving on to the next section of the code.

Setting the value to `1` means only one process at a time. This will use the least amount of memory, but will result in the longest compile time since we are only compiling for one process at a time. Increasing the value increases the total amount of memory, but should decrease the compile time.

If you are only worried about functional testing in a memory constrained environment and not on compile time efficiency then setting this value to `1` is likely what you want.

**Comment by thanh-lam** [comment]
@jjhursey , thanks! This is very helpful.
Any estimate when or will these new parameters be added to `inference.py`?
Because this "helps" get around the model failure: "Resource temporarily unavailable" with 8-AIU tests, we need to document it while waiting for "real" memory fixes.

**Comment by jjhursey** [comment]
Both the x86 and Power test teams have confirmed that this helps mitigate their testing needs for large models on low memory systems.

**Comment by jjhursey** [comment]
@JRosenkranz @ani300 this is ready for review

File: aiu_fms_testing_utils/utils/__init__.py [hunk 1]
```diff
@@ -3,26 +3,42 @@
...
+def warmup_model(model: nn.Module, input_ids: torch.Tensor, max_new_tokens: int, compile_dynamic_sendnn = False, stagger_update_lazyhandle = 0, **padding_kwargs):
     import torch_sendnn
     dprint("AIU warmup")
-    pt_compile_model_time = time.time()
     extra_kwargs = {**padding_kwargs, "only_last_token": True}
     max_new_tokens_warmup = max_new_tokens
     if compile_dynamic_sendnn:
         max_new_tokens_warmup = 2
+
+    if stagger_update_lazyhandle > 0 and stagger_update_lazyhandle != world_size:
```
Comment (JRosenkranz) [inline aiu_fms_testing_utils/utils/__init__.py:22]: It looks like logic is called multiple times, do you think it would make sense to put it in it's own utility function, this way in can be re-used in the future

File: scripts/inference.py [hunk 2]
```diff
@@ -218,6 +220,24 @@
     default=0,
     help="Set verbosity level (pass flag as `-v`, `-vv`, `-vvv`)"
 )
+parser.add_argument(
+    "--stagger_load",
+    type=int,
+    default=0,
+    help="Stagger model loading to avoid OOM issues on the host"
```
Comment (JRosenkranz) [inline scripts/inference.py:227]: can you provide an example of what setting this value does

File: scripts/inference.py [hunk 3]
```diff
@@ -218,6 +220,24 @@
...
+)
+parser.add_argument(
+    "--stagger_update_lazyhandle",
+    type=int,
+    default=0,
+    help="Stagger update_lazyhandle to avoid OOM issues on the host"
```
Comment (JRosenkranz) [inline scripts/inference.py:233]: can you provide an example of what setting this value does

**Comment by dpatel-ops** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.2-8b-instruct BATCH_SIZE=8 SEQUENCE_LENGTH=64 USE_TINY_MODEL=0 NUM_AIU=2

File: scripts/inference.py (hunk 2 above)
Comment (jjhursey) [reply scripts/inference.py:227]: When you launch a 16 way TP model and set `--stagger_load 2` then it'll only allow up to `2` processes to enter the `get_model` phase of the code at a time instead of all 16.

How about this for a rewording of the help text?
```suggestion
    help="Limit the number of concurrent processes executing the model loading phase. Set to 0 to allow all processes"
```

File: scripts/inference.py (hunk 3 above)
Comment (jjhursey) [reply scripts/inference.py:233]: When you launch a 16 way TP model and set `--stagger_update_lazyhandle 2` then it'll only allow up to `2` processes to enter the `update_lazyhandle ` phase of the code at a time instead of all 16.

How about this for a rewording of the help text?
```suggestion
    help="Limit the number of concurrent processes executing the AIU update_lazyhandle phase. Set to 0 to allow all processes"
```

**Comment by jjhursey** [comment]
I pushed a commit to consolidate the staggered enter/exit. I also rebased on main

File: aiu_fms_testing_utils/utils/__init__.py [hunk 4]
```diff
@@ -3,26 +3,48 @@
...
+from aiu_fms_testing_utils.utils.aiu_setup import dprint, rank, local_rank, world_size
 from typing import Optional, List, Tuple
 import os
 import requests
 import json
 import random
+import math
 
-def warmup_model(model: nn.Module, input_ids: torch.Tensor, max_new_tokens: int, compile_dynamic_sendnn = False, **padding_kwargs):
+def stagger_enter(limit: int):
```
Comment (JRosenkranz) [inline aiu_fms_testing_utils/utils/__init__.py:14]: nit: would you mind  just adding a single docstring for `stagger_enter` and `stagger_leave`. We will need to add one to `warmup_model` as well, but doesn't need to be from this PR

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.2-8b-instruct BATCH_SIZE=8 SEQUENCE_LENGTH=64 USE_TINY_MODEL=0

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.2-8b-instruct BATCH_SIZE=8 SEQUENCE_LENGTH=64 USE_TINY_MODEL=0

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.2-8b-instruct BATCH_SIZE=8 SEQUENCE_LENGTH=64 USE_TINY_MODEL=0

File: aiu_fms_testing_utils/utils/__init__.py [hunk 5]
```diff
@@ -3,26 +3,62 @@
...
+    :param limit: Number of concurrent processes allowed in the code section if > 0.
+    """
+    if limit > 0 and limit != world_size:
+        for _set in range( math.ceil(world_size / float(limit)) ):
+            if rank < (_set+1)*limit:
+                break
+            torch.distributed.barrier()
+        dprint(f"Stagger: Enter (Set: {_set+1} of {math.ceil(world_size / float(limit))})")
+
+def stagger_leave(limit: int):
```
Comment (JRosenkranz) [inline aiu_fms_testing_utils/utils/__init__.py:29]: Would it make sense for this to be a `stagger_context` (given each stagger_enter needs to be paired with a stagger_leave):

```python
with stagger_context():
     model = get_model(...)
```

Comment (jjhursey) [reply aiu_fms_testing_utils/utils/__init__.py]: I added a commit to convert this to a contextlib function. Take a look and let me know what you think.

**Comment by jjhursey** [comment]
I recently updated my foundation-model-stack repo, and now I'm seeing a hang during the `generate` wrapped by `with torch_sendnn.warmup_mode():` because the second iteration is calling allreduce. So the allreduce is hanging because not all of the processes are participating in the warmup at the same time.

**Comment by austinpagan** [comment]
So, I created a container from a recent image provided by my team lead.
I cloned aiu-fms-testing-utils.
I created a branch in that clone, containing this PR (called "test47_branch"), using this command:

`git fetch origin pull/47/head:test47_branch`

I ran a test using the main branch of the clone, and the test ran to completion.

Then, I followed these steps inside aiu-fms-testing-utils to install this PR in the container, using the branch I created:

```
pip3 uninstall aiu-fms-testing-utils
git checkout test47_branch
pip3 install -e .
```

Then I ran my test again, NOT specifying any stagger_update_lazyhandle value, and the test terminated about six minutes after the compile_graph binary started running.
It displayed the following error message:

```
DtException: dynamicDsg->hasNode(kv.first), file /project_src/deeptools/dsm/graphOptimizer2.cpp line 1515
[DeepRT] ======= dscGlobal.showOpt() =====
[DeepRT] ======= Dumping 1024 sengraph(s) to /tmp/spyre-ffdc-2025-07-28-21-58-50-75045-r1-4 =====
[DeepRT] ======= DtException CAUGHT =====
```

If there is some other step or steps I should have taken, I welcome any advice that you have.

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_scripts.py

**Comment by JRosenkranz** [comment]
bot:test
TEST_FILE=test_decoders.py MODEL_ID=ibm-granite/granite-3.3-8b-instruct BATCH_SIZE=8 SEQUENCE_LENGTH=64 USE_TINY_MODEL=0
